src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -870,6 +894,34 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 
 
 /*
+ * GetPlacementConnectionsReadyForTwoPhaseCommit iterates over the task placement list,
+ * get connections and mark them as critical. Finally, establishes connections to all
+ * placements at once.
+ */
+static void
+GetPlacementConnectionsReadyForTwoPhaseCommit(List *taskPlacementList)
+{
+	ListCell *taskPlacementCell = NULL;
+	List *multiConnectionList = NIL;
+
+	foreach(taskPlacementCell, taskPlacementList)
+	{
+		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = GetNodeConnection(connectionFlags,
```

Comment:
This doesn't establish connections at once? For that you'd need to use StartNodeConnection() for all of them, and then in a second loop finish the establishment.